### PR TITLE
*: add_date can return mysql.Time

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

to #9813 

### What is changed and how it works?

Implement things according to MySQL source code.

### Check List 

Tests 

 - Unit test (still adding more)
 - Integration test(will add, if unit test is hard to cover some cases)

Side effects

 - Increased code complexity

Related changes

 - Need to be included in the release note

### CODE REVIEW

**Comment by winoros** [comment]
The overflow checking is not finished. And i think the previous add_date will cause overflow too...
I'm not sure whether to fix them together in next pr or fix this one first and leave the original add_date part to next pr...

File: types/const_test.go [hunk 1]
```diff
@@ -30,11 +29,6 @@ import (
 	"github.com/pingcap/tidb/util/testleak"
 )
 
-func TestT(t *testing.T) {
```
Comment (winoros) [inline types/const_test.go]: It's a redundant one.

File: types/time.go [hunk 2]
```diff
@@ -1820,10 +1829,115 @@ func ExtractTimeValue(unit string, format string) (int64, int64, int64, float64,
...
 }
 
+func extractSingleDurationValue(unit string, format string) (Duration, error) {
+	fv, err := strconv.ParseFloat(format, 64)
+	if err != nil {
+		return ZeroDuration, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
+	}
+	iv := int64(math.Round(fv))
+
+	switch strings.ToUpper(unit) {
```
Comment (tiancaiamao) [inline types/time.go]: Redundant computing as `ExtractDurationValue` already call the `strings.ToUpper` function

File: types/time.go [hunk 3]
```diff
@@ -1820,10 +1829,115 @@ func ExtractTimeValue(unit string, format string) (int64, int64, int64, float64,
...
+			return ZeroDuration, err
+		}
+		return Duration{Duration: gotime.Duration(day*int64(GoDurationDay) + int64(dur)), Fsp: 0}, nil
+	case "YEAR_MONTH":
+		_, _, _, _, err := extractYearMonth(format)
+		if err != nil {
+			return ZeroDuration, err
+		}
+		// MONTH must exceed the limit of mysql's duration. So just return overflow error.
+		return ZeroDuration, ErrDatetimeFunctionOverflow.GenWithStackByArgs("time")
```
[the remark this answers is not in the document]
Comment (winoros) [reply types/time.go]: Oh, my concern is that though it must exceed the limit, should we raise the overflow error when the content cannot be formatted in fact?

File: types/time.go [hunk 4]
```diff
@@ -1820,10 +1827,115 @@ func ExtractTimeValue(unit string, format string) (int64, int64, int64, float64,
...
 		return extractYearMonth(format)
 	default:
-		return 0, 0, 0, 0, errors.Errorf("invalid singel timeunit - %s", unit)
+		return 0, 0, 0, 0, errors.Errorf("invalid single timeunit - %s", unit)
 	}
 }
 
+// ExtractDurationValue extract the value from format to Duration.
+func ExtractDurationValue(unit string, format string) (Duration, error) {
+	switch strings.ToUpper(unit) {
```
Comment (tiancaiamao) [inline types/time.go]: I mean here you can

```
unit = strings.ToUpper(unit)
switch unit {
 ...
}
```

Make sure that `MICROSEconD` `SeConD` all work likewise (case insensitive) in mysql

Comment (winoros) [reply types/time.go]: yes it can work.

**Comment by zz-jason** [comment]
@winoros Please update this pr branch with the latest master

File: expression/builtin_time.go [hunk 5]
```diff
@@ -2735,6 +2735,18 @@ func (du *baseDateArithmitical) add(ctx sessionctx.Context, date types.Time, int
 	return date, false, nil
 }
 
+func (du baseDateArithmitical) addDuration(ctx sessionctx.Context, d types.Duration, interval string, unit string) (types.Duration, bool, error) {
```
Comment (qw4990) [inline expression/builtin_time.go]: Should be `*baseDateArithmitical`?

File: types/time.go [hunk 6]
```diff
@@ -1737,10 +1732,116 @@ func ExtractTimeValue(unit string, format string) (int64, int64, int64, float64,
...
 }
 
+// ExtractDurationValue extract the value from format to Duration.
+func ExtractDurationValue(unit string, format string) (Duration, error) {
+	unit = strings.ToUpper(unit)
+	switch unit {
+	case "MICROSECOND", "SECOND", "MINUTE", "HOUR", "DAY", "WEEK", "MONTH", "QUARTER", "YEAR":
+		return extractSingleDurationValue(unit, format)
+	case "SECOND_MICROSECOND":
+		_, _, _, d, err := extractTimeValue(format, MicrosecondIndex, SecondMicrosecondMaxCnt)
```
Comment (qw4990) [inline types/time.go]: After this PR(#10329), large `duration` will be rounded up to `day`, so it's better to keep and check all unit fields.

File: types/time.go [hunk 7]
```diff
@@ -1737,10 +1732,116 @@ func ExtractTimeValue(unit string, format string) (int64, int64, int64, float64,
...
+	}
+	iv := int64(math.Round(fv))
+
+	switch unit {
+	case "MICROSECOND":
+		return Duration{Duration: gotime.Duration(fv * float64(gotime.Microsecond)), Fsp: MaxFsp}, nil
+	case "SECOND":
+		return Duration{Duration: gotime.Duration(fv * float64(gotime.Second)), Fsp: 0}, nil
+	case "MINUTE":
+		return Duration{Duration: gotime.Duration(fv * float64(gotime.Minute)), Fsp: 0}, nil
```
Comment (qw4990) [inline types/time.go]: Like issue #9727, `fv*gotime.Minute` may cause int64 overflow.

File: types/time.go (hunk 6 above)
Comment (winoros) [reply types/time.go]: i'll change codes when that one merged.

**Comment by winoros** [comment]
Wait for #10329 merged first.

**Comment by tiancaiamao** [comment]
LGTM

File: types/time.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zz-jason) [reply types/time.go]: is it controlled by a certain sql mode?

**Pull request merged**